doc: Add non-thread-safe note to FeeFilterRounder::round()

hebasto opened the pull request:
The `FastRandomContext` class is documented as not thread-safe.
This PR adds a relevant note to the `FeeFilterRounder::round()` function declaration.

Close #19254

MarcoFalke commented on src/policy/fees.h:
```diff
@@ -281,11 +281,12 @@ class FeeFilterRounder
     explicit FeeFilterRounder(const CFeeRate& minIncrementalFee);
 
     /** Quantize a minimum fee for privacy purpose before broadcast **/
```
```suggestion
    /** Quantize a minimum fee for privacy purpose before broadcast . Not thread-safe due to use of FastRandomContext **/
```

sipa replied:
If this is not going to be changed, it's probably better to make it an explicit global.

MarcoFalke replied:
Oh, I am not suggestion to never change this. This can be changed when net_processing is multi-threaded. Until then documentation and sanitizers are good enough to leave the code as is for now.

ajtowns replied:
Isn't it as thread safe as a `class A { int a = 0;  public: int increment() { return ++a; } }`, which we'd solve by saying `static A my_a GUARDED_BY(cs_main);` ? And both could be made parallelisable by saying `static thread_local A my_a;` -- it's just that `GUARDED_BY` doesn't work with block-scope statics?

Just adding `// protected by cs_main` at the declaration of filterRounder, or moving it to be a module-level static and adding a `GUARDED_BY(cs_main)` seems better to me?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19268#issuecomment-653873471).

maflcko reviewed (commented):
My preference would be to just document it. Unless this is the only random context without a lock held in the net_processing loop.

As soon as the net_processing loop is run in parallel, our sanitizers should point out any unsafe code, so there should be no risk in leaving the code as-is.

naumenkogs commented:
Concept ACK.

I don't have a strong opinion on taking either approach (mutex vs documentation-only)

ajtowns commented on src/net_processing.cpp:
```diff
@@ -4396,7 +4396,7 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
             int64_t timeNow = GetTimeMicros();
             if (timeNow > pto->m_tx_relay->nextSendTimeFeeFilter) {
                 static CFeeRate default_feerate(DEFAULT_MIN_RELAY_TX_FEE);
```
Can make `default_feerate` const too -- could make it `constexpr` even, so long as the constructors were also marked constexpr.

hebasto committed:
doc: Add non-thread-safe note to FeeFilterRounder::round()

Co-authored-by: MarcoFalke <falke.marco@gmail.com>

hebasto commented:
Updated 877cca538b29bbb7420601e7972fb5635523f507 -> d842e6ac965b528f0d704f54aceb91eae84085fb ([pr19268.01](https://github.com/hebasto/bitcoin/commits/pr19268.01) -> [pr19268.02](https://github.com/hebasto/bitcoin/commits/pr19268.02)):

- rebased
- implemented @MarcoFalke's [suggestion](https://github.com/bitcoin/bitcoin/pull/19268#discussion_r439748886)

MarcoFalke commented:
self ACK d842e6ac965b528f0d704f54aceb91eae84085fb

practicalswift commented:
ACK d842e6ac965b528f0d704f54aceb91eae84085fb: explicit is better than implicit

naumenkogs commented:
ACK d842e6a
